packages/entity-server/src/routes/hierarchy/relations/types.ts

```diff
@@ -18,27 +20,47 @@ export type Prefix<T, P extends string> = {
   [field in keyof T & string as `${P}_${field}`]: T[field];
 };
 
-export type RelationsSubQuery = Omit<HierarchyRequestQuery, 'fields'>;
+export type RelationsSubQuery = Omit<SingleEntityRequestQuery, 'fields'>;
 export type RelationsQuery = RelationsSubQuery & {
   groupBy?: 'ancestor' | 'descendant';
 } & Partial<Prefix<RelationsSubQuery, 'ancestor'>> &
   Partial<Prefix<RelationsSubQuery, 'descendant'>>;
 
-type RelationsSubContext = {
+type DescendantSubContext = {
   filter: EntityFilter;
   field: keyof FlattableEntityFields;
   type: string;
 };
+
+type AncestorSubContext = {
+  filter: EntityFilter;
+  field: keyof FlattableEntityFields;
+  type?: string;
+};
+
+export type RelationsResponseBody =
+  | Record<FlattenedEntity, EntityResponse> // groupBy: descendant
+  | Record<FlattenedEntity, EntityResponse[]>; // groupBy: ancestor
+
+export type RelationsContext = Omit<SingleEntityContext, 'fields'> & {
+  ancestor: AncestorSubContext;
+  descendant: DescendantSubContext;
+};
+
 export interface RelationsRequest
+  extends Request<SingleEntityRequestParams, RelationsResponseBody, RequestBody, RelationsQuery> {
+  ctx: RelationsContext;
+}
+
+export interface MultiEntityRelationsRequest
   extends Request<
-    HierarchyRequestParams,
-    | Record<FlattenedEntity, EntityResponse> // groupBy: descendant
-    | Record<FlattenedEntity, EntityResponse[]>, // groupBy: ancestor
-    HierarchyRequestBody,
-    RelationsQuery
+    MultiEntityRequestParams,
+    RelationsResponseBody,
+    RequestBody,
+    RelationsQuery & { entities?: string }
```

Comment:
Are `entities` here optional?

Reply:
Technically yes. All URL query parameters are optional in the sense that the user's request will still hit the middleware/route even if they didn't add the parameter. I actually like making all query parameters optional in typescript for this reason, as it forces my middleware/route to do validation/error handling for the case that the user forgot to add the right parameters 😄

We set the `entities` on the `MultiEntityContext` as non-optional since by the time we are setting the context we've ensured the url parameter is present 👍